### PR TITLE
rootfs: Add support for rootfs-overlay.

### PR DESCRIPTION
**Pull request by flouthoc**
Allows users to specify a readonly rootfs with `:O`, in exchange podman will create a writable overlay.

Add support for `podman run --rootfs /some/readonly/path:O echo hello`.
Depends Upon: https://github.com/containers/buildah/pull/3434

Idea is to re-use `graphRoot` ( i.e  `<base>`+ `containers/storage/`+ `overlay-containers` ) to sustain overlay-ed `rootfs`.

Closes: https://github.com/containers/podman/issues/11038

### CODE REVIEW

**Comment by flouthoc** [comment]
wait for https://github.com/containers/buildah/pull/3434

**Comment by flouthoc** [comment]
@rhatdan Latest buildah sync is still needed for this PR can we do it without cutting a new buildah release ?

**Comment by rhatdan** [comment]
Yes, just update go.mod with the buildah vendor ID, and we can get this merged.  We only need new version when we are preparing a release.

**Comment by flouthoc** [comment]
buildah-bud tests are having conflicts i am checking this. @rhatdan other than that PR is ready for review. Could you take a look.

PS: Its a single commit as `build commit` test would fail without latest buildah vendor.

**Comment by flouthoc** [comment]
Updating [`test/buildah-bud/buildah-tests.diff`](https://github.com/containers/podman/pull/11170/files#diff-ce24690bf576e8daa82507aa1ef275a12224dc3a89005736c95293e86d72bedfL1)

**Comment by flouthoc** [comment]
@rhatdan @giuseppe PTAL. We can also put buildah vendor sync in a different PR but if its in the same pr changes should be in same commit to pass `build commit test`.

File: libpod/options.go
Comment (rhatdan) [inline libpod/options.go:1382]: Return rootfs here.

Comment (rhatdan) [reply libpod/options.go:1382]: Who is supposed to unmount the overlay.Mount?

Comment (flouthoc) [reply libpod/options.go:1382]: @rhatdan mount path will be removed  as soon as the  container is deleted since we are mounting at `<base(graphroot)>+ containers/storage/+ overlay-containers`. But I can explicit clean up if that is needed.

Comment (rhatdan) [reply libpod/options.go:1382]: If it gets cleaned up, then fine. I just want  you to make sure.

mount | wc
podman run ... echo hi
mount | wc

Comment (flouthoc) [reply libpod/options.go:1382]: @rhatdan verified with `mount | wc` and rootless mode cant see any dangling mounts

Comment (flouthoc) [reply libpod/options.go:1382]: @rhatdan for the first comment,  `"", nil` is returned so caller will automatically check if string is empty it will fallback to original rootfs.

Comment (flouthoc) [reply libpod/options.go:1382]: @rhatdan PTAL

Comment (rhatdan) [reply libpod/options.go:1382]: If you always return rootfs, then the caller will not need to check.

Comment (flouthoc) [reply libpod/options.go:1382]: Sure let me fix this.

Comment (flouthoc) [reply libpod/options.go]: @rhatdan fixed in latest commit.

**Comment by edsantiago** [comment]
I see the problem; it's from https://github.com/containers/buildah/pull/3409/files#diff-0ca52cf2d14756ee50e1b99c5b46dfc1c45a1a796746ebe6a6f406055dba929e which I never reviewed.

As best I can tell, your new `buildah-tests.diff` looks good; and the `bud` tests are passing, which is further confirmation. Does that address your concern? If not, could you remind me of your question?

**Comment by flouthoc** [comment]
@edsantiago this solves my concern thanks

File: docs/source/markdown/podman-build.1.md
Comment (rhatdan) [inline docs/source/markdown/podman-build.1.md:543]: Should this be --stdin?

**Review by rhatdan — APPROVED** [review]
Nice new feature. Too bad you don't document it so that people would know about it.

How about you add some information to a man page. :^)

File: docs/source/markdown/podman-build.1.md
Comment (flouthoc) [reply docs/source/markdown/podman-build.1.md:543]: @rhatdan no this was added in buildah but not added in `man pages` CI was failing because of this. Since command was there on `podman build` but not in man pages I added this manually.

Comment (rhatdan) [reply docs/source/markdown/podman-build.1.md:543]: Well I am confused.  Heading is --ssh and then comment talks about --stdin.  No meantion of --ssh?

**Comment by rhatdan** [comment]
You need to document this feature in the podman run and podman create man page, along with examples.

File: docs/source/markdown/podman-build.1.md
Comment (flouthoc) [reply docs/source/markdown/podman-build.1.md:543]: @rhatdan its coming from this PR https://github.com/containers/buildah/pull/3409 because of buildah sync, since new build option was added to buildah but podman would also needs man page to be updated , following `--ssh` in unrelated to current PR.

**Comment by flouthoc** [comment]
@rhatdan Added to `man-page` with examples.

File: docs/source/markdown/podman-build.1.md
Comment (rhatdan) [reply docs/source/markdown/podman-build.1.md:543]: I see this in buildah bud man page?

```
       --ssh=default|id[=socket>|<key>[,<key>]

       SSH agent socket or keys to expose to the build.  The socket path can be left empty to use the value of default=$SSH_AUTH_SOCK

       To later use the ssh agent, use the --mount flag in a RUN instruction within a Containerfile:

```

Comment (rhatdan) [reply docs/source/markdown/podman-build.1.md:543]: **--ssh**=**default**|*id[=socket>|<key>[,<key>]*

SSH agent socket or keys to expose to the build.
The socket path can be left empty to use the value of `default=$SSH_AUTH_SOCK`

To later use the ssh agent, use the --mount flag in a `RUN` instruction within a `Containerfile`:

`RUN --mount=type=secret,id=id mycmd`

File: docs/source/markdown/podman-create.1.md
Comment (rhatdan) [inline docs/source/markdown/podman-create.1.md:862]: What happens if I try to `podman commit` one of these images?

**Comment by rhatdan** [comment]
LGTM other then the --ssh man page.

File: docs/source/markdown/podman-create.1.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-create.1.md:857]:
```suggestion
can modify content within the mount point which is stored in the
```

File: docs/source/markdown/podman-create.1.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-create.1.md:855]: Going back and forth over keeping as is or removing the 'a'....  your call.
```suggestion
   The `:O` flag tells Podman to mount the directory from the rootfs path as
```

File: docs/source/markdown/podman-run.1.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-run.1.md:895]: changes from above here too

File: docs/source/markdown/podman-create.1.md
Comment (flouthoc) [reply docs/source/markdown/podman-create.1.md]: @TomSweeneyRedHat resolved in latest commit.

File: docs/source/markdown/podman-run.1.md
Comment (flouthoc) [reply docs/source/markdown/podman-run.1.md:895]: @TomSweeneyRedHat resolved in latest commit.

**Comment by flouthoc** [comment]
This is weird dont know why tests are failing after rebase i am checking this.

**Comment by rhatdan** [comment]
LGTM

**Comment by TomSweeneyRedHat** [comment]
@flouthoc you need to rebase and your tests are all an unhappy red.

**Comment by flouthoc** [comment]
It should pass this time. :sweat_smile:

**Comment by flouthoc** [comment]
@rhatdan @TomSweeneyRedHat PTAL

**Comment by rhatdan** [comment]
@vrothberg @baude @giuseppe @nalind @TomSweeneyRedHat @mheon @containers/podman-maintainers PTAL

File: libpod/options.go
Comment (baude) [inline libpod/options.go:1380]: would it make sense to reverse the conditional logic to avoid a return the if condition?

File: libpod/options.go
Comment (lsm5) [inline libpod/options.go:1362]: is this comment still relevant?

Comment (baude) [inline libpod/options.go:1362]: is this still true?

File: docs/source/markdown/podman-build.1.md
Comment (giuseppe) [inline docs/source/markdown/podman-build.1.md:543]: could we move this chunk to a separate commit as well as a separate commit for the buildah version bump?

**Comment by baude** [comment]
couple of comments for consideration ... else LGTM

File: libpod/options.go
Comment (ashley-cui) [inline libpod/options.go:1362]: Does this comment still apply?

File: libpod/options.go
Comment (mheon) [inline libpod/options.go:1341]: Can we move the parsing higher in the stack? This would be a lot cleaner if it was `WithRootFS(rootfs string, overlay bool) ...`

File: libpod/options.go
Comment (mheon) [inline libpod/options.go:1373]: This doesn't seem right - are we keeping the overlay mounted continuously? Can't we mount it only when the container is running? What happens if the system reboots and the FS is unmounted?

Comment (flouthoc) [reply libpod/options.go:1373]: @mheon It is there as long as the container exists since there could be cases were stuff is written on overlay mount that was the original idea to keep mount intact and let it cleaned up as soon as container is removed.

File: docs/source/markdown/podman-build.1.md
Comment (flouthoc) [reply docs/source/markdown/podman-build.1.md:543]: @giuseppe `Build commit test` fails since original commit is dependent upon updated builah. We could create a seperate PR for version bump and merge that first.

File: libpod/options.go
Comment (mheon) [reply libpod/options.go:1373]: That doesn't sound right - we don't like to leave mounts around, taking up resources on the host system, when containers are not running. It seems more correct to move this into container mount / unmount logic, where it will act like other containers do.

File: docs/source/markdown/podman-build.1.md
Comment (rhatdan) [reply docs/source/markdown/podman-build.1.md:550]: @ashley-cui if this is true, could you open PR against BUILDAH to make same change.

**Comment by flouthoc** [comment]
I think we can create a seperate PR for buildah vendoring and changes. It'll be easier to mange.

File: docs/source/markdown/podman-build.1.md
Comment (flouthoc) [reply docs/source/markdown/podman-build.1.md:550]: @ashley-cui Could you create a PR in buildah, I'll remove buildah changes and vendoring from this PR.

File: libpod/options.go
Comment (flouthoc) [reply libpod/options.go:1373]: @mheon Inoder to expect rootfs as overlay this has to be mounted before container is invoked. This might not follow the path of regular mounts but I'll add changes to perform cleanups.

Comment (mheon) [reply libpod/options.go:1373]: Sure - every container mounts before starting, as part of the normal mount process. We should be able to hook the existing `mountStorage()` function to mount the overlay. Or is there some other limitations here I'm not aware of?

**Comment by rhatdan** [comment]
@flouthoc What is going on with this PR?

**Comment by flouthoc** [comment]
@rhatdan small hardening needs to done here  i'll do that in a day or two and will move out buildah vendoring changes in a different PR so this is easier to review and merge.

**Comment by flouthoc** [comment]
This is rebased and hardend approach.

File: libpod/options.go
Comment (flouthoc) [reply libpod/options.go]: this is resolved.

Comment (flouthoc) [reply libpod/options.go]: this is resolved.

File: libpod/options.go
Comment (flouthoc) [reply libpod/options.go]: @mheon this is resolved.

File: libpod/options.go
Comment (flouthoc) [reply libpod/options.go]: this is resolved and moved to `mountStorage(`

File: docs/source/markdown/podman-build.1.md
Comment (flouthoc) [reply docs/source/markdown/podman-build.1.md]: @ashley-cui I have fixed this in current PR could you please create a PR for buildah if its needed.

**Review by rhatdan — APPROVED** [review]
LGTM

**Comment by rhatdan** [comment]
I think it is in podman 3.4 release.

**Comment by flouthoc** [comment]
@vrothberg Ackd @giuseppe reported this few days back. Cleanups are leaking i'll push a fix with https://github.com/containers/podman/pull/11937

**Pull request merged**
